src/pallene/coder.lua

```diff
@@ -245,17 +245,23 @@ function Coder:get_stack_slot(typ, dst, slot, loc, description_fmt, ...)
     else
         assert(not typ.is_upvalue_box)
         local extra_args = table.pack(...)
+        local expected_type = pallene_type_tag(typ)
+        if expected_type == "LUA_TUSERDATA" then
+            expected_type = C.string(typ.name)
+        else
+            expected_type = "pallene_tag_name(" .. expected_type .. ")"
+        end
         check_tag = util.render([[
             if (l_unlikely(!$test)) {
                 pallene_runtime_tag_check_error(L,
-                    $file, $line, $expected_tag, rawtt($slot),
+                    $file, $line, $expected_type, pallene_type_name(L, $slot),
```

Comment:
Perhaps it would be more efficient to pass the `$slot` directly and let the `pallene_runtime_tag_check_error` compute the type name. That way, we don't need to generate code for the type name function call at run-time tag check.